File: pkg/scope/governance.go
package scope

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
)

const (
	SubjectTypePlatform = "platform"
	SubjectTypeMerchant = "merchant"

	DefaultPlatformID int64 = 1
)

type GovernanceScope struct {
	ScopeType  string `json:"scopeType"`
	PlatformID int64  `json:"platformId"`
	TenantID   int64  `json:"tenantId,omitempty"`
	MerchantID int64  `json:"merchantId,omitempty"`
}

type GovernanceScopeMetadata struct {
	GovernanceScope
	ActivationStatus string `json:"activationStatus,omitempty"`
	RoleMode         string `json:"roleMode,omitempty"`
	ScopeLabel       string `json:"scopeLabel,omitempty"`
}

func NormalizeGovernanceScope(scopeType string, platformID, tenantID, merchantID int64) (GovernanceScope, error) {
	normalized := GovernanceScope{
		ScopeType:  strings.TrimSpace(scopeType),
		PlatformID: platformID,
		TenantID:   tenantID,
		MerchantID: merchantID,
	}
	if normalized.ScopeType == "" {
		normalized.ScopeType = inferScopeType(tenantID, merchantID)
	}

	if normalized.PlatformID == 0 {
		normalized.PlatformID = DefaultPlatformID
	}

	switch normalized.ScopeType {
	case SubjectTypePlatform:
		normalized.TenantID = 0
		normalized.MerchantID = 0
	case SubjectTypeTenant:
		if normalized.TenantID <= 0 {
			return GovernanceScope{}, errors.New("租户级治理必须提供 tenantId")
		}
		normalized.MerchantID = 0
	case SubjectTypeMerchant:
		if normalized.MerchantID <= 0 {
			return GovernanceScope{}, errors.New("商户级治理必须提供 merchantId")
		}
	default:
		return GovernanceScope{}, fmt.Errorf("不支持的主体范围：%s", normalized.ScopeType)
	}

	return normalized, nil
}

func inferScopeType(tenantID, merchantID int64) string {
	if merchantID > 0 {
		return SubjectTypeMerchant
	}
	if tenantID > 0 {
		return SubjectTypeTenant
	}
	return SubjectTypePlatform
}

func (s GovernanceScope) Label() string {
	switch s.ScopeType {
	case SubjectTypeTenant:
		return fmt.Sprintf("租户 #%d", s.TenantID)
	case SubjectTypeMerchant:
		if s.TenantID > 0 {
			return fmt.Sprintf("租户 #%d / 商户 #%d", s.TenantID, s.MerchantID)
		}
		return fmt.Sprintf("商户 #%d", s.MerchantID)
	default:
		return "平台级"
	}
}

func (s GovernanceScope) Key() string {
	return fmt.Sprintf("%s:%d:%d:%d", s.ScopeType, s.PlatformID, s.TenantID, s.MerchantID)
}

func (s GovernanceScope) SameScope(other GovernanceScope) bool {
	return s.ScopeType == other.ScopeType &&
		s.PlatformID == other.PlatformID &&
		s.TenantID == other.TenantID &&
		s.MerchantID == other.MerchantID
}

func EncodeGovernanceScopeMetadata(scope GovernanceScope, activationStatus, roleMode string) (string, error) {
	payload, err := json.Marshal(GovernanceScopeMetadata{
		GovernanceScope:  scope,
		ActivationStatus: strings.TrimSpace(activationStatus),
		RoleMode:         strings.TrimSpace(roleMode),
		ScopeLabel:       scope.Label(),
	})
	if err != nil {
		return "", err
	}

	return string(payload), nil
}

File: web-admin/src/pages/system/components/governance.ts
export type GovernanceScopeType = 'platform' | 'tenant' | 'merchant';

export interface GovernanceScopeValue {
  scopeType?: GovernanceScopeType;
  scopeLabel?: string;
  platformId?: number;
  tenantId?: number;
  merchantId?: number;
}

export const DEFAULT_PLATFORM_ID = 1;

export const defaultGovernanceScope: GovernanceScopeValue = {
  scopeType: 'platform',
  platformId: DEFAULT_PLATFORM_ID,
  tenantId: 0,
  merchantId: 0,
  scopeLabel: '平台级 #1',
};

export function buildGovernanceScopeLabel(input?: Partial<GovernanceScopeValue>): string {
  const scope = {
    scopeType: (input?.scopeType || 'platform') as GovernanceScopeType,
    platformId: input?.platformId || DEFAULT_PLATFORM_ID,
    tenantId: input?.tenantId || 0,
    merchantId: input?.merchantId || 0,
  };

  if (scope.scopeType === 'tenant') {
    return scope.tenantId ? `租户级 #${scope.tenantId}` : '租户级';
  }
  if (scope.scopeType === 'merchant') {
    return scope.merchantId ? `商户级 #${scope.merchantId}` : '商户级';
  }
  return `平台级 #${scope.platformId}`;
}

export const normalizeGovernanceScope = (
  input?: Partial<GovernanceScopeValue>,
): GovernanceScopeValue => {
  const scopeType = (input?.scopeType || 'platform') as GovernanceScopeType;
  const platformId =
    input?.platformId && input.platformId > 0 ? input.platformId : DEFAULT_PLATFORM_ID;
  const tenantId = scopeType === 'platform' ? 0 : Number(input?.tenantId || 0);
  const merchantId = scopeType === 'merchant' ? Number(input?.merchantId || 0) : 0;

  return {
    scopeType,
    platformId,
    tenantId,
    merchantId,
    scopeLabel: buildGovernanceScopeLabel({
      scopeType,
      platformId,
      tenantId,
      merchantId,
    }),
  };
};

export const toGovernancePayload = (input?: Partial<GovernanceScopeValue>) => {
  const scope = normalizeGovernanceScope(input);
  return {
    scopeType: scope.scopeType,
    platformId: scope.platformId,
    tenantId: scope.tenantId,
    merchantId: scope.merchantId,
  };
};

export const governanceScopeColor = (scopeType?: GovernanceScopeType) => {
  switch (scopeType) {
    case 'tenant':
      return 'green';
    case 'merchant':
      return 'orange';
    default:
      return 'blue';
  }
};

export const governanceImpactText = (
  input: Partial<GovernanceScopeValue>,
  entityLabel = '治理元数据',
) => {
  const scope = normalizeGovernanceScope(input);
  if (scope.scopeType === 'tenant') {
    return `当前只会影响该租户下的${entityLabel}、表单选项和通知配置，不会串写平台级默认数据。`;
  }
  if (scope.scopeType === 'merchant') {
    return `当前只会影响目标商户的${entityLabel}。如需平台或租户复用，请切回更高层级后再操作。`;
  }
  return `当前维护的是平台默认${entityLabel}，后续租户与商户后台可直接复用。`;
};
